### PR TITLE
JSON serialization of Realm.Object & Realm.Collection

### PR DESCRIPTION
**Pull request by steffenagger**
#### TS declarations (used in integration-tests)
- [x] Limit input in `create<T>(...)` to match a model (using @nirinchev's `ExtractPropertyNamesOfType`©) **notion of any implications this could introduce, is welcome!**.
- [x] Allow input to contain `Array<T>` in `RealmPartialModel<T>` instead of a defined `Realm.Collection<T>` in `create<T>(...)`.
- [x] Update `objects<T>(...)` to accommodate the underlying logic & different responses.
- [x] Allow extending `Realm.Object` for class models.

#### `toJSON` on `Realm.Object` & `Realm.Collection` (this includes `Realm.List` & `Realm.Results`)
- [x] Implement solution with object references (this introduces a cache to not return new objects).
- [x] Ensure `TypeError: Converting circular structure to JSON` is thrown for circular structures, not `RangeError: Maximum call stack size exceeded`.
- [x] Implement & expose replacer for circular structures. \*

\* _I've included a combined key of the table-type & `_objectId()` as unique `$refId` in `Realm.JsonSerializationReplacer`, for a (visual) reference in serialized JSON. This is only used in case a primaryKey is not defined._

#### Integration Tests
- [x] Write test covering `toJSON` implementations for both interface/schema & class model usage.

#### Changelog
- [x] Update `CHANGELOG.md`

### CODE REVIEW

File: lib/extensions.js
Comment (kraenhansen) [inline lib/extensions.js:144]: Probably nit picking here, but it would make it a bit more DRY.
```suggestion
            Object.keys(this)
                .concat(Object.keys(Object.getPrototypeOf(this)));
                .forEach(k => moveAsSerializable(cache, k, this, result));
```

File: lib/extensions.js
Comment (kraenhansen) [inline lib/extensions.js:795]:
```suggestion
                        seen.splice(pos + 1);
```

File: lib/extensions.js
Comment (kraenhansen) [inline lib/extensions.js:797]:
```suggestion
                        seen.push(parent);
```

File: lib/extensions.js
Comment (kraenhansen) [inline lib/extensions.js:804]:
```suggestion
                    seen.push(value);
```

😉

File: lib/extensions.js
Comment (kraenhansen) [inline lib/extensions.js:797]:
```suggestion
                        seen.push(parent);
```

File: lib/extensions.js
Comment (kraenhansen) [inline lib/extensions.js:804]:
```suggestion
                    seen.push(value);
```

File: lib/index.d.ts
Comment (kraenhansen) [inline lib/index.d.ts:833]: I for one (obviously 🤔 ) know what it does - but for future maintainers of the code, we might need a small explanation of this piece of black magic.

File: lib/index.d.ts
Comment (kraenhansen) [inline lib/index.d.ts:835]: I don't know if `Model` is the right word here - we're already using `ObjectSchema` and `Class` elsewhere.

Comment (nirinchev) [inline lib/index.d.ts:835]: I expect to see license fees once that gets merged ;p

Comment (steffenagger) [reply lib/index.d.ts:835]: Of course! And I imagine they'll just go directly out of my salary... 😁

File: lib/extensions.js
Comment (nirinchev) [inline lib/extensions.js:805]: This section of the code definitely needs comments - I'm having a very hard time wrapping my head around it and confirming it's behaving as expected, especially considering it's modifying some external array.

File: integration-tests/tests/src/serialization.ts
Comment (kraenhansen) [inline integration-tests/tests/src/serialization.ts:51]: It would be great with a text which explicitly tests the behaviour `toJSON` with circular references.
Specifically I would expect `john` to be the same object (by reference) if he is befriended with himself (or via other person).

File: lib/extensions.js
Comment (steffenagger) [reply lib/extensions.js:138]: True. Those were moved, thanks.

File: lib/extensions.js
Comment (steffenagger) [reply lib/extensions.js:805]: Yes definitely, it's on the todo - maybe the draft/WIP was opened a bit early, just to show the overall approach.

File: lib/extensions.js
Comment (steffenagger) [reply lib/extensions.js]: I think we should add `"semi": [1, "always"]` to the eslint config then.

File: lib/extensions.js
Comment (nirinchev) [reply lib/extensions.js:805]: The PR description was longer than the PR itself, so of course I completely skipped it 😅

File: lib/extensions.js
Comment (steffenagger) [reply lib/extensions.js:144]: I like it. It also opens for having `moveAsSerializable` back in as an anonymous function.

File: lib/extensions.js
Comment (steffenagger) [reply lib/extensions.js:126]: Oh... good to know, and back to the drawing board. Think I have a solution though.

File: lib/extensions.js
Comment (steffenagger) [reply lib/extensions.js]: It's not.

File: lib/extensions.js
Comment (steffenagger) [reply lib/extensions.js]: Wouldn't the new `const id = this.constructor.name + this._objectId();` fix this?

File: integration-tests/tests/src/serialization.ts
Comment (steffenagger) [reply integration-tests/tests/src/serialization.ts:73]: included in the updated tests.

File: lib/extensions.js
Comment (steffenagger) [reply lib/extensions.js]: Added comments.

**Comment by blagoev** [comment]
Is this a draft still

File: lib/index.d.ts
Comment (steffenagger) [reply lib/index.d.ts:835]: Comments added

**Review by blagoev — APPROVED** [review]

File: lib/index.d.ts
Comment (steffenagger) [reply lib/index.d.ts]: This has now changed name to `RealmInsertionModel` I personally think it's quite descriptive and a commonly used name. I don't the either Schema or Class fits here, but could use `RealmInsertionObject`?
This naming is already in v10, so I suggest we revisit the naming if it's confusing to users?

File: integration-tests/tests/src/serialization.ts
Comment (kneth) [inline integration-tests/tests/src/serialization.ts:1]: Please add copyright header.

**Review by kneth — APPROVED** [review]

**Comment by KalAvroniev** [comment]
Still getting `RangeError: Maximum call stack size exceeded`.
It happens on records with this schema when using toJSON
`{
  "name": "tag",
  "primaryKey": "uuid",
  "properties": {
    "tags": {
      "type": "list",
      "objectType": "tag"
    }
  }
}`

**Pull request merged**
